src/handlers/browser/recorders/cdp/page.ts

```diff
@@ -0,0 +1,192 @@
+    })
+
+    this.#client.page.on('frameStartedNavigating', ({ data }) => {
+      if (data.frameId !== this.#id) {
+        return
+      }
+
+      this.#startedNavigation = data
+    })
+
+    this.#client.page.on('frameNavigated', ({ data }) => {
+      if (data.frame.id !== this.#id) {
+        return
+      }
+
+      // Ignore navigations caused by something happening with the page (user interaction, script, etc)
+      if (this.#requestedNavigation !== null) {
+        return
+      }
+
+      if (this.#startedNavigation === null) {
+        logger.warn(
+          'Received frameNavigated event without prior navigation events'
+        )
+
+        return
+      }
+
+      if (isReload(this.#startedNavigation)) {
+        this.emit('navigate', {
+          event: {
+            type: 'reload-page',
+            eventId: uuid(),
+            timestamp: Date.now(),
+            tab: this.#id,
+            url: data.frame.url,
+          },
+        })
+
+        this.#reset()
+
+        return
+      }
+
+      const source = toNavigationSource(this.#startedNavigation)
+
+      if (source === null) {
+        this.#reset()
+
+        return
+      }
+
+      this.emit('navigate', {
+        event: {
+          type: 'navigate-to-page',
+          eventId: uuid(),
+          timestamp: Date.now(),
+          source,
+          url: data.frame.url,
+          tab: this.#id,
+        },
+      })
+
+      this.#reset()
+    })
+
+    this.#client.page.on('frameStoppedLoading', ({ data }) => {
+      if (data.frameId !== this.#id) {
+        return
+      }
+
+      this.#reset()
+    })
+
+    this.#script.on('reload', () => {
+      this.#client.page.reload({}).catch((error) => {
+        logger.error('Failed to reload page:', error)
+      })
+    })
+  }
+
+  async attach() {
+    await this.#client.page.enable()
+    await this.#client.page.setBypassCSP(true)
+
+    await this.#client.page.addScriptToEvaluateOnNewDocument({
+      source: `window.__K6_STUDIO_TAB_ID__ = "${this.#id}";`,
+      runImmediately: true,
+    })
+
+    await this.#script.inject(this.#client)
+
+    await this.#client.runtime.runIfWaitingForDebugger()
+
+    return this
+  }
+
+  navigateTo(url: string) {
+    this.#client.page
+      .navigate({ url, transitionType: 'other' })
+      .catch((error) => {
+        logger.error('Failed to navigate page:', error)
+      })
+  }
+
+  #reset() {
+    this.#requestedNavigation = null
+    this.#startedNavigation = null
+  }
+
+  /**
+   * Convencience method to log page events for debugging purposes
+   */
+  #trace() {
+    const events: Array<keyof CdpPage.EventMap> = [
+      'frameRequestedNavigation',
+      'frameStartedNavigating',
+      'frameNavigated',
+      'frameStartedLoading',
+      'frameStoppedLoading',
+    ]
+
+    for (const eventName of events) {
+      this.#client.page.on(eventName, ({ data }) => {
+        console.log(eventName, data)
+      })
+    }
+  }
```

Comment:
The behavior of these events are poorly documented so I've been using this to try and figure out which event fires when. I'm leaving it to aid in future debugging.